base/utf.jl

```diff
@@ -0,0 +1,969 @@
+    @inbounds return UTF16String(setindex!(copy!(Vector{T}(len), dat), 0, len))
+end
+@inline function fast_utf_copy(T::Type{Char}, len, dat)
+    @inbounds return UTF32String(setindex!(copy!(Vector{T}(len), dat), 0, len))
+end
+
+# Get rest of character ch from 3-byte UTF-8 sequence in str
+@inline function get_utf8_3(dat, pos, ch)
+    @inbounds return ((ch & 0xf) << 12) | (UInt32(dat[pos-1] & 0x3f) << 6) | (dat[pos] & 0x3f)
+end
+
+# Get rest of character ch from 4-byte UTF-8 sequence in dat, update pos and return character
+@inline function get_utf8_4(dat, pos, ch)
+    @inbounds return ((ch & 0x7) << 18) | (UInt32(dat[pos-2] & 0x3f) << 12) | (UInt32(dat[pos-1] & 0x3f) << 6) | (dat[pos] & 0x3f)
+end
+
+# Output a character as a 4-byte UTF-8 sequence, update the position
+@inline function output_utf8_4(buf, out, ch)
+    @inbounds begin
+        buf[out + 1] = 0xf0 | (ch >>> 18)
+        buf[out + 2] = 0x80 | ((ch >>> 12) & 0x3f)
+        buf[out + 3] = 0x80 | ((ch >>> 6) & 0x3f)
+        buf[out + 4] = 0x80 | (ch & 0x3f)
+    end
+end
+
+#=
+"""
+@brief      Converts an AbstractString to a UTF16String
+
+@param[in]  ::Type{UTF16String}
+@param[in]  str::AbstractString
+
+@return     ::UTF16String
+@throws     ArgumentError
+"""
+=#
+function convert(::Type{UTF16String}, str::AbstractString)
+    len, flags, num4byte = check_string_abs(str)
+    buf = Vector{UInt16}(len+num4byte+1)
+    out = 0
+    @inbounds for ch in str
+        c = reinterpret(UInt32, ch)
+        if c < 0x10000
+            buf[out += 1] = UInt16(c)
+        else
+            # output surrogate pair
+            buf[out += 1] = UInt16(0xd7c0 + (ch >>> 10))
+            buf[out += 1] = UInt16(0xdc00 + (ch & 0x3ff))
+        end
+    end
+    @inbounds buf[out + 1] = 0 # NULL termination
+    UTF16String(buf)
+end
+
+#=
+"""
+@brief      Converts an AbstractString to a UTF32String
+
+@param[in]  ::Type{UTF32String}
+@param[in]  str::AbstractString
+
+@return     ::UTF32String
+@throws     ArgumentError
+"""
+=#
+function convert(::Type{UTF32String}, str::AbstractString)
+    len, flags = check_string_abs(str)
+    buf = Vector{Char}(len+1)
+    out = 0
+    @inbounds for ch in str ; buf[out += 1] = ch ; end
+    @inbounds buf[out + 1] = 0 # NULL termination
+    UTF32String(buf)
+end
+
+#=
+@doc """
+@brief      Converts a UTF-8 encoded string to UTF-16 encoding
+
+@param[in]  str::Vector{UInt8}
+
+@return     ::UTF16String
+@throws     ArgumentError
+""" ->
+=#
+function convert(::Type{UTF16String}, str::UTF8String)
+    dat = str.data
+    # handle zero length string quickly
+    sizeof(dat) == 0 && return empty_utf16
+    # Check that is correct UTF-8 encoding and get number of words needed
+    len, flags, num4byte = check_string_utf8(dat)
+    len += num4byte
+    buf = Vector{UInt16}(len+1)
+    buf[len+1] = 0
+    # Optimize case where no characters > 0x7f
+    flags == 0 && @inbounds return UTF16String(copy!(buf, dat))
```

Comment:
All occurrences of `@inbounds return` basically (though some may be valid, not sure).